python/taichi/main.py

```diff
@@ -487,26 +170,604 @@ def main(debug=False):
+        parser.add_argument('-v',
+                            '--verbose',
+                            dest='verbose',
+                            action='store_true',
+                            help='Run with verbose outputs')
+        parser.add_argument('-r',
+                            '--rerun',
+                            required=False,
+                            default=None,
+                            dest='rerun',
+                            type=str,
+                            help='Rerun failed tests for given times')
+        parser.add_argument(
+            '-t',
+            '--threads',
+            required=False,
+            default=None,
+            dest='threads',
+            type=str,
+            help='Custom number of threads for parallel testing')
+        args = parser.parse_args(arguments)
+
+        # Short circuit for testing
+        if self.test_mode: return args
+
+        if args.files:
+            if args.cpp:
+                return TaichiMain._test_cpp(args)
+            else:
+                return TaichiMain._test_python(args)
+        elif args.cpp:
+            # Only run C++ tests
+            return TaichiMain._test_cpp(args)
+        else:
+            # Run both C++ and Python tests
+            ret = TaichiMain._test_python(args)
+            if ret != 0:
+                return ret
+            return TaichiMain._test_cpp(args)
+
+    def debug(self, arguments: list = sys.argv[2:]):
+        """Debug a single script"""
+        parser = argparse.ArgumentParser(description=f"{self.debug.__doc__}")
+        parser.add_argument(
+            'filename',
+            help='A single (Python) script to run with debugger, e.g. render.py'
+        )
+        args = parser.parse_args(arguments)
+
+        # Short circuit for testing
+        if self.test_mode: return args
+
+        ti.core.set_core_trigger_gdb_when_crash(True)
+
+        with open(args.filename) as script:
+            script = script.read()
+
+        # FIXME: exec is a security risk here!
+        exec(script, {'__name__': '__main__'})
+
+    def run(self, arguments: list = sys.argv[2:]):
+        """Run a specific task"""
+        # TODO: discuss do we really need this?
+        parser = argparse.ArgumentParser(description=f"{self.run.__doc__}")
+        parser.add_argument('taskname',
+                            help='A single task name to run, e.g. test_math')
+        parser.add_argument('taskargs',
+                            nargs='*',
+                            help='Optional task argument(s) to run with task')
+        args = parser.parse_args(arguments)
+
+        # Short circuit for testing
+        if self.test_mode: return args
+
+        task = ti.Task(args.taskname)
+        task.run(*args.taskargs)
+
+
+def main():
+    TaichiMain()
 
 
 def main_debug():
-    main(debug=True)
+    TaichiMain(debug=True)
 
 
-if __name__ == '__main__':
-    exit(main())
+if __name__ == "__main__":
+    main()
```

Comment:
We probably need to `exit` with the return value here. This is important `ti test`: if any test fails, the CLI should return `1` instead of `0`. It seems to me that we are always returning `0` now.

Reply:
Thanks, this is a really great catch. I haven't thought about the return codes that much and have been treating the commands as side-effects here. Do we have other commands that works similar to `ti test`, on top of your mind? Asking since I'm struggling between 2 very different approaches:

**A**
```diff
def test(self, arguments: list = sys.argv[2:]):
    if args.files:
        if args.cpp:
-            return TaichiMain._test_cpp(args)
+            exit(TaichiMain._test_cpp(args))
        else:
-            return TaichiMain._test_python(args)
+            exit(TaichiMain._test_python(args))
    elif args.cpp:
        # Only run C++ tests
-       return TaichiMain._test_cpp(args)
+       exit(TaichiMain._test_cpp(args))
    else:
        # Run both C++ and Python tests
        ret = TaichiMain._test_python(args)
        if ret != 0:
-           return ret
+           exit(ret)
-       return TaichiMain._test_cpp(args)
+       exit(TaichiMain._test_cpp(args))
```

**B**
```diff

def __init__(self, debug: bool = False, test_mode: bool = False):
-    getattr(self, args.command)(sys.argv[2:])

+def __call__(self):
+    getattr(self, args.command)(sys.argv[2:])

def main():
-  TaichiMain()
+  cli = TaichiMain()
+  cli()
```

Apparently the latter one is more portable, but requires changing the tests and the main() rapper to always be there around the TaichiMain class.

Reply:
Thanks for providing the nice options here. I also think B is better. Having too many `exits()` in the program makes the control flow less clear.

Reply:
Cool, lemme flush this out asap.